README.md
MeggyJrRGB
==========

Software for the Meggy Jr RGB kit from Evil Mad Scientist Laboratories.

The main documentation page for the Meggy Jr RGB kit is here: 
http://wiki.evilmadscientist.com/Meggy_Jr_RGB



This repository is new, and should be considered to be supplementary to that at our old google code site, here: https://code.google.com/p/meggy-jr-rgb/


---------------------------------


Meggy Jr RGB is a platform to develop handheld pixel games based around an 8x8 RGB LED matrix display. The heart of the system is an ATmega168 or ATmega328P microcontroller,which can be programmed through the Arduino development environment.


Meggy Jr RGB is an open source project designed and released by Evil Mad Scientist Laboratories. (Open Source Hardware & Open Source Software!)

For more information, please see the Meggy Jr RGB project page and the introduction to programming Meggy Jr RGB at our wiki: http://wiki.evilmadscientist.com/Meggy_Jr_RGB


We also have a developer's mailing list for Meggy Jr RGB; if you have questions, that might be a good place too: https://groups.google.com/forum/#!forum/meggydev

Links to additional Meggy Jr RGB programs can be found at the Evil Mad Scientist wiki.

Changes.txt
v 1.5  Modifications for compatibility with Arduino 1.0, and new RGB LED Matrix displays with brighter green LEDs.

 - Example files renamed with ".ino" file extension (suffix).

 - Requirement for wiring.h removed from MeggyJr.cpp file.

 - Added new color map choice, in the MeggyJr.h file.  For newer LED matrices, with stronger green elements, leave the default choice (#define UseNewColors  1) enabled.  For older matrices-- or if your white looks purple --comment out that line.

 - Changes to serial sketches, to reflect changes in Arduino 1.0.





v 1.4: Added new SafeDrawPx to the simplified library.
  -This function is like DrawPx, except that it performs limit checking on the inputs.

Also added new program, MeggyJr_Revenge.


v 1.3.2:
Re-zipped to (hopefully) avoid issues with the Zip format on Vista.


Upgrade note for library version 1.3

The sound generation routine has been changed, and sounds now terminate
automatically.  The functions SoundCheck() and Tone_Update() are no longer needed and should be removed from programs where they were previously found.

The variables ToneInProgress and SoundStopTime have also been removed.  To determine if a sound is currently playing you can check to see if the value of MakingSound is nonzero.

Example:      
    Tone_Start(2000, 50); 
    while (MakingSound)   {}    


Upgrade note for library version 1.2

If moving from an older version of the library, please note that the following function two names have changed:

1. 'EmptyScreen' is now 'ClearSlate'

and

2.  ShowScreen  is now 'DisplaySlate'

The syntax and functionality of these two functions has not changed, however the names have been changed to make the documentation more clear.  If you have written software for Meggy Jr that uses these functions, a simple search-and-replace operation should bring you up to full compatibility with library version 1.2.  We'll try not to do this again. :)

MeggyJr.cpp
/*
  MeggyJr.cpp - Meggy Jr RGB library for Arduino
 Version 1.5 - 12/31/2011
 Copyright (c) 2011 Windell H. Oskay.  All right reserved.
 
    This library is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this library.  If not, see <http://www.gnu.org/licenses/>.
	
	Thanks to Arthur J. Dahm III and Jay Clegg for code written for the Peggy 2.0,
	which was adapted to make this library.  
	
*/

/******************************************************************************
 * Includes
 ******************************************************************************/
 
#include "MeggyJr.h" 

byte MeggyJr::MeggyFrame[DISP_BUFFER_SIZE];

byte MeggyJr::AuxLEDs;
byte MeggyJr::SoundAllowed;	 
byte MeggyJr::SoundEnabled;	

byte MeggyJr::currentCol;
byte MeggyJr::currentBrightness;
byte* MeggyJr::currentColPtr;
		 
//unsigned long MeggyJr::SoundStopTime;	

unsigned int MeggyJr::ToneTimeRemaining;

#ifdef timingtest
unsigned int MeggyJr::testTime;
#endif
	
/******************************************************************************
 * Constructor
 ******************************************************************************/

MeggyJr::MeggyJr(void)			
{
// Initialization routine for Meggy Jr RGB hardware

	AuxLEDs = 0;
	currentColPtr = MeggyFrame;
	currentCol=0;
	currentBrightness=0;
	
PORTC = 255U;	// Pull-ups on Port C  (for detecting button presses)
DDRC = 0;		//All inputs
  
DDRD = 254U;		// All D Output except for Rx
PORTD = 254U;	
		
DDRB = 63U;	
PORTB = 255;		

//If a button is pressed at startup, turn sound off.
 if ((PINC & 63) != 63)
{
  SoundAllowed = 0;
} 
else
{
  SoundAllowed = 1;
} 
   
   SoundEnabled = 0;
   
MeggyJr::ToneTimeRemaining = 0;
   
//Turn display off:
PORTD |= 252U;
PORTB |= 17U;

ClearMeggy();
  	 
SPSR = (1<<SPI2X); 
/*
//ENABLE SPI, MASTER, CLOCK RATE fck/4:	
SPCR = 80;// i.e., (1 << SPE) | ( 1 << MSTR ); 
 
SPDR = 0;
 while (!(SPSR & (1<<SPIF)))  { } // wait for last bitshift to complete
SPDR = 0;
 while (!(SPSR & (1<<SPIF)))  { } // wait for last bitshift to complete
SPDR = 0;
 while (!(SPSR & (1<<SPIF)))  { } // wait for last bitshift to complete
SPDR = 0;
 while (!(SPSR & (1<<SPIF)))  { } // wait for last bitshift to complete
 
PORTB |= 4;		//Latch Pulse    
PORTB &= 251;
*/
SPCR = 0; //turn off spi

// setup the interrupt.
TCCR2A = (1<<WGM21); // clear timer on compare match
TCCR2B = (1<<CS21); // timer uses main system clock with 1/8 prescale
OCR2A  = (F_CPU >> 3) / 8 / 15 / FPS; // Frames per second * 15 passes for brightness * 8 rows
TIMSK2 = (1<<OCIE2A);	// call interrupt on output compare match

sei( );    // Enable interrupts
	
	#ifdef timingtest
MeggyJr::testTime = 0;
#endif
	
	
}

/******************************************************************************
 * User API
 ******************************************************************************/
 
// Painfully Slow!  Don't use this, if you can avoid it:
void MeggyJr::ClearMeggy (void)
{
	for (byte i = 0; i < DISP_BUFFER_SIZE; i++)
	{ 
	    MeggyFrame[i]= 0;
	}
}
 
//Set Pixel Color:  use an RGB array to specify the color.
// Very convenient for using color look-up tables.
void MeggyJr::SetPxClr(byte x, byte y, byte *rgb)
{
  byte PixelPtr =  24*x + y;
  MeggyFrame[PixelPtr] = rgb[2];   
  PixelPtr += 8;
  MeggyFrame[PixelPtr] = rgb[1];
  PixelPtr += 8;
  MeggyFrame[PixelPtr] = rgb[0]; 
}
  
 
  
/* 
void MeggyJr::SoundCheck(void)   
  { 
 // Obsolete with current version of library; sounds stop automatically.
 // If your program contains "SoundCheck();" somewhere, please remove it.
}
*/  
   

// Begin sound 
void  MeggyJr::StartTone(unsigned int Tone, unsigned int duration)   
  {
  OCR1A = Tone;
  SoundState(1);
  MeggyJr::ToneTimeRemaining = duration;  
  }
    	
 
byte MeggyJr::GetPixelR(byte x, byte y)
{
  return MeggyFrame[24*x + y + 16]; 
}

byte MeggyJr::GetPixelG(byte x, byte y)
{
  return MeggyFrame[24*x + y + 8]; 
}

byte MeggyJr::GetPixelB(byte x, byte y)
{
  return MeggyFrame[24*x + y]; 
}

// Clear a single pixel.  Not much better than writing "dark" to the pixel.
void MeggyJr::ClearPixel(byte x, byte y)
{
byte PixelPtr =  24*x + y;
MeggyFrame[PixelPtr] = 0;   
PixelPtr += 8;
MeggyFrame[PixelPtr] = 0;
PixelPtr += 8;
MeggyFrame[PixelPtr] = 0; 
}

 
// GetButtons returns a byte with a bit set for 
// each of the buttons that is currently pressed.

byte MeggyJr::GetButtons(void)
{
  return (~(PINC) & 63U); 
}



// Set sound ON or OFF by calling
// SoundState(0) or SoundState(1).

void MeggyJr::SoundState(byte t)
{

if ((t) && (SoundAllowed))
	{ 
    
 		TCCR1A = 65;	// 0100 0001 
		//COM1A10 = 01 :  Toggle output OC1A on compare match.
		//WGM11,10: 01

		TCCR1B = 17;	// 0001 0001  	// CS12..10 = 001: No clock prescaling. 
		// WGM13,12: 10
		// CS12,CS11,CS10: 001.  Count at CLKI/O/1 (no prescaling)

		//WGM: 1001, phase+frequency correct PWM
		// with top at OCR1A and update OCR1A at bottom.
		
        SoundEnabled = 1;
		DDRB |= 2; 
	}
else
	{
	    SoundEnabled = 0;
		TCCR1A = 0;	

	if (t)
	    TCCR1B = 128;		// Harmless; can use (TCCR1B == 0) to check if sound is done.
	else
		TCCR1B = 0; 	 
		DDRB &= 253;
		PORTB |= 2;
	}
}
  

SIGNAL(TIMER2_COMPA_vect)
{			
	// there are 15 passes through this interrupt for each row per frame.
	// ( 15 * 8) = 120 times per frame.
	// during those 15 passes, a led can be on or off.
	// if it is off the entire time, the perceived brightness is 0/15
	// if it is on the entire time, the perceived brightness is 15/15
	// giving a total of 16 average brightness levels from fully off to fully on.
	// currentBrightness is a comparison variable, used to determine if a certain
	// pixel is on or off during one of those 15 cycles.   
	//
	// At 120 Hz refresh, this executes 15*8*120 = 14,400 times per second.
	// Full routine takes about 340 cycles, 21 us. 
	//
	// 14400*340/16000000 = ~0.3; so about 1/3 of the processor time is spent in 
	// this interrupt routine.

#ifdef timingtest
unsigned int soundTemp = OCR1A;
OCR1A = 65500;
TCNT1 = 0;
#endif

	if (++MeggyJr::currentBrightness >= MAX_BRIGHTNESS)  
	{
		MeggyJr::currentBrightness = 0;
		if (++MeggyJr::currentCol > 7)
		{
			MeggyJr::currentCol = 0; 
			MeggyJr::currentColPtr = MeggyJr::MeggyFrame;
		}
		else
			MeggyJr::currentColPtr += 24;
			 
		if (MeggyJr::ToneTimeRemaining >  0)
			{ 
				if (--MeggyJr::ToneTimeRemaining == 0)
					{
						TCCR1A = 0;	  
						DDRB &= 253;
						PORTB |= 2;
						TCCR1B = 0;  // High bit flags that the sound is done.
					} 
			} 
	}
		
	////////////////////  Parse a row of data and write out the bits via spi
  
	byte *ptr = MeggyJr::currentColPtr + 23;  // it is more convenient to work from right to left
	byte p;
	byte cb =  MeggyJr::currentBrightness; 
	// Optimization: interleave waiting for SPI with other code, so the CPU can do something useful
	// when waiting for each SPI transmission to complete

//Turn display off:
PORTD |= 252U;
PORTB |= 17U;
 
SPCR = 80;// i.e., (1 << SPE) | ( 1 << MSTR );   

// First SPI word:  Aux LED drive.  Zero, except once per full screen redraw.	 
  
if ((cb + MeggyJr::currentCol ) == 0)
	SPDR = MeggyJr::AuxLEDs;
else
    SPDR = 0; 
        
	byte bits=0; 
 
	p = *ptr--;
	if (p > cb)  			
          bits |= 128;
	p = *ptr--;
	if (p > cb)  			
          bits |= 64;
	p = *ptr--;
	if (p > cb)  			
          bits |= 32;
	p = *ptr--;
	if (p > cb)  				
          bits |= 16;
	p = *ptr--;
	if (p > cb)  				
          bits |= 8;
	p = *ptr--;
	if (p > cb)  				
          bits |= 4;
	p = *ptr--;
	if (p > cb)  				
          bits |= 2;
	p = *ptr--;
	if (p > cb)  				
          bits |= 1;
		    
//	while (!(SPSR & (1<<SPIF)))  { } // wait for prior bitshift to complete
	SPDR = bits;
	
	bits=0;
	p = *ptr--;
	if (p > cb)  				
          bits |= 128;
	p = *ptr--;
	if (p > cb)  				
          bits |= 64;
	p = *ptr--;
	if (p > cb)  				
          bits |= 32;
	p = *ptr--;
	if (p > cb)  				
          bits |= 16;
	p = *ptr--;
	if (p > cb)  				
          bits |= 8;
	p = *ptr--;
	if (p > cb)  				
          bits |= 4;
	p = *ptr--;
	if (p > cb)  				
          bits |= 2;
	p = *ptr--;
	if (p > cb)  				
          bits |= 1;

//	while (!(SPSR & (1<<SPIF)))  { } // wait for prior bitshift to complete
	SPDR = bits;
		
	bits=0;
	p = *ptr--;
	if (p > cb)  				
          bits |= 128;
	p = *ptr--;
	if (p > cb)  				
          bits |= 64;
	p = *ptr--;
	if (p > cb)  				
          bits |= 32;
	p = *ptr--;
	if (p > cb)  				
          bits |= 16;
	p = *ptr--;
	if (p > cb)  				
          bits |= 8;
	p = *ptr--;
	if (p > cb)  				
          bits |= 4;
	p = *ptr--;
	if (p > cb)  				
          bits |= 2;
	p = *ptr--;
	if (p > cb)  				
          bits |= 1;

//asm("nop");	// short delay


	while (!(SPSR & (1<<SPIF)))  { } // wait for prior bitshift to complete
	SPDR = bits;
	
	////////////////////  Now set the row and latch the bits
	byte portbTemp = 0;	
	byte portdTemp = 0;

	if (MeggyJr::currentCol == 0)
		 portbTemp = 239U;
	else if (MeggyJr::currentCol == 1)
		 portbTemp = 254U;
	else
		 portdTemp = ~(1 << (9 - MeggyJr::currentCol));
   
	while (!(SPSR & (1<<SPIF)))  { } // wait for last bitshift to complete
 
PORTB |= 4;		//Latch Pulse    

    if (MeggyJr::currentCol > 1)
       PORTD &= portdTemp;
    else
       PORTB &= portbTemp;

       PORTB &= 251;           //End Latch Pulse
       SPCR = 0; //turn off spi 
	    
#ifdef timingtest

if (TCNT1 > MeggyJr::testTime)
   MeggyJr::testTime = TCNT1;
   
   OCR1A =  soundTemp;    
#endif
	
}



MeggyJrSimple.h
/*
  MeggyJrSimple.h 
   The Meggy Jr Simplified Library (MJSL)
   
 Part of the Meggy Jr RGB library for Arduino
 Version 1.5 - 12/31/2011
 Copyright (c) 2011 Windell H. Oskay.  All right reserved.
 
 This library is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this library.  If not, see <http://www.gnu.org/licenses/>.
 
 Thanks to Arthur J. Dahm III and Jay Clegg for code written for the Peggy 2.0,
 which was adapted to make this library.  
 
 */



#ifndef MeggyJrSimple_h
#define MeggyJrSimple_h
 
#include <MeggyJr.h>

MeggyJr Meg;
byte GameSlate[8][8];       
byte lastButtonState;

byte Button_A;		 
byte Button_B;
byte Button_Up;
byte Button_Down;
byte Button_Left;
byte Button_Right;
 
#define MeggyCursorColor   15,15,15				// You can define color constants like this.

//Color lookup Table
byte ColorTable[26][3] = 
{
  { MeggyDark  },  
  { MeggyRed  }  ,
  { MeggyOrange  },
  { MeggyYellow  },
  { MeggyGreen  },
  { MeggyBlue  } ,
  { MeggyViolet  },
  { MeggyWhite  },  
  { MeggyDimRed  },  
  { MeggyDimOrange  },  
  { MeggydimYellow  },  
  { MeggyDimGreen  },  
  { MeggydimAqua  },  
  { MeggyDimBlue  },  
  { MeggydimViolet  },   
  { MeggyCursorColor},      //Extra bright cursor position color (not white).
  {0,0,0},					//CustomColor0 (dark, by default)
  {0,0,0},			        //CustomColor1 (dark, by default)
  {0,0,0},					//CustomColor2 (dark, by default)
  {0,0,0},					//CustomColor3 (dark, by default)	
  {0,0,0},					//CustomColor4 (dark, by default)
  {0,0,0},					//CustomColor5 (dark, by default)
  {0,0,0},					//CustomColor6 (dark, by default)
  {0,0,0},					//CustomColor7 (dark, by default)
  {0,0,0},					//CustomColor8 (dark, by default)
  {0,0,0}					//CustomColor9 (dark, by default)
}; 

// Assign those colors names that we can use:
enum colors {
  Dark, Red, Orange, Yellow, Green, Blue, Violet, White, 
  DimRed,DimOrange,DimYellow,DimGreen,DimAqua,DimBlue,DimViolet,FullOn,
    CustomColor0, CustomColor1, CustomColor2, CustomColor3, CustomColor4,  
	CustomColor5, CustomColor6, CustomColor7, CustomColor8, CustomColor9
	};     


  
void CheckButtonsDown()
	{ 
	 byte i = Meg.GetButtons(); 
	 
 	 Button_B  = (i & 1);      
     Button_A = (i & 2);     
     Button_Up = (i & 4);
     Button_Down = (i & 8);
     Button_Left = (i & 16);
     Button_Right = (i & 32);
	 
	 lastButtonState = i; 
	}
	 
void CheckButtonsPress()
	{
	 byte j;
	 byte i = Meg.GetButtons();
	 j = i & ~(lastButtonState);  // What's changed?
	 
 	 Button_B  = (j & 1);      
     Button_A = (j & 2);     
     Button_Up = (j & 4);
     Button_Down = (j & 8);
     Button_Left = (j & 16);
     Button_Right = (j & 32);
	 
	 lastButtonState = i;
	}
 
 
// Write a byte to the Auxiliary LED set at the top of the LED matrix display.  
void SetAuxLEDs(byte InputLEDs)
	{
		Meg.AuxLEDs = InputLEDs;
	}

 

// Write a byte to the Auxiliary LED set at the top of the LED matrix display.  
// This version reverses bit order, so you can call it with an explicit binary number
void SetAuxLEDsBinary(byte n)
{
n = (n & 240) >> 4 | (n & 15) << 4; 
n = (n & 204) >> 2 | (n & 51) << 2; 
Meg.AuxLEDs = (n & 170) >>1 | (n & 85) << 1; 
}

 


// Simple function to color in a pixel at position (x,y,color):
void DrawPx(byte xin, byte yin, byte color)
{
		GameSlate[xin][yin] = color;
}


// Same as above, except checks to see if pixel is onscreen
// This function is new as of v 1.4
void SafeDrawPx(byte xin, byte yin, byte color)
{
    if ((xin >= 0) && (xin <= 7) && (yin >= 0) && (yin <= 7))
		GameSlate[xin][yin] = color;
}

// function to read color of pixel at position (x,y):
byte ReadPx(byte xin, byte yin)
{   
	return	GameSlate[xin][yin];
}


//Empty the Game Slate:
void ClearSlate(void)
{
  byte i;
  byte j;
  i = 0;
  while (i < 8) { 
    j = 0;
    while ( j < 8)
    {
      GameSlate[i][j] = 0;
      j++;
    }
    i++;
  }
}


// DisplaySlate() :: Write the Game Slate to the Display Memory it.
// This function looks up each color number (name) stored in the Game Slate,
// retreives its R,G,B components from the color table, and writes them to the
// R,G,B parts of the display memory.

void DisplaySlate (void) {	
  byte  j = 0; 
  while (j < 8) 
  {
    Meg.SetPxClr(j, 7, ColorTable[ GameSlate[j][7] ]);   
    Meg.SetPxClr(j, 6, ColorTable[ GameSlate[j][6] ]);  
    Meg.SetPxClr(j, 5, ColorTable[ GameSlate[j][5] ]);   
    Meg.SetPxClr(j, 4, ColorTable[ GameSlate[j][4] ]);  
    Meg.SetPxClr(j, 3, ColorTable[ GameSlate[j][3] ]);   
    Meg.SetPxClr(j, 2, ColorTable[ GameSlate[j][2] ]);  
    Meg.SetPxClr(j, 1, ColorTable[ GameSlate[j][1] ]);   
    Meg.SetPxClr(j, 0, ColorTable[ GameSlate[j][0] ]);  
    j++; 
  }  	 
}  
 #endif
  
 
  // Pre-defined sound divisors:
 
#define ToneB2		64783 

#define ToneC3		61157	
#define ToneCs3		57724
#define ToneD3		54485	
#define ToneDs3		51427
#define ToneE3		48541	
#define ToneF3		45816	
#define ToneFs3		43243
#define ToneG3		40816	
#define ToneGs3		38526
#define ToneA3		36363
#define ToneAs3		34323
#define ToneB3		32397
 
#define ToneC4		30578
#define ToneCs4		28862
#define ToneD4		27242
#define ToneDs4		25713
#define ToneE4		24270
#define ToneF4		22908
#define ToneFs4		21622
#define ToneG4		20408
#define ToneGs4		19263
#define ToneA4		18182 
#define ToneAs4		17161
#define ToneB4		16198 

#define ToneC5		15289
#define ToneCs5		14431
#define ToneD5		13626
#define ToneDs5		12857
#define ToneE5		12135
#define ToneF5		11454
#define ToneFs5		10811
#define ToneG5		10204
#define ToneGs5		9631
#define ToneA5		9091 
#define ToneAs5		8581
#define ToneB5		8099 

#define ToneC6		7645
#define ToneCs6		7215
#define ToneD6		6810
#define ToneDs6		6428
#define ToneE6		6067
#define ToneF6		5727
#define ToneFs6		5405
#define ToneG6		5102
#define ToneGs6		4816
#define ToneA6		4545 
#define ToneAs6		4290
#define ToneB6		4050 

#define ToneC7		3822
#define ToneCs7		3608
#define ToneD7		3406
#define ToneDs7		3214
#define ToneE7		3034
#define ToneF7		2863
#define ToneFs7		2703
#define ToneG7		2551
#define ToneGs7		2408
#define ToneA7		2273 
#define ToneAs7		2145
#define ToneB7		2025 

#define ToneC8		1911
#define ToneCs8		1804
#define ToneD8		1703
#define ToneDs8		1607
#define ToneE8		1517
#define ToneF8		1432
#define ToneFs8		1351
#define ToneG8		1276
#define ToneGs8		1204
#define ToneA8		1136 
#define ToneAs8		1073
#define ToneB8		1012 

#define ToneC9		956
#define ToneCs9		902
#define ToneD9		851
#define ToneDs9		803

  /*
 Sound output:
 
 Frequency is given by the following formula:
 
 f= 8 MHz/divisor, so divisor = 8 MHz/f.   (Round to nearest.)
 Maximum divisor: 65535, so min. frequency is 122 Hz
 
 Example: for 440 Hz, divisor = 18182
  
 */
  
   // "Cheater" functions
 #define SoundOn()    Meg.SoundState(1)
 #define SoundOff()   Meg.SoundState(0) 
 #define MakingSound  (TCCR1B > 0)		
   
 
 
 
 // other sound functions:
void Tone_Start(unsigned int divisor, unsigned int duration_ms)
{
  Meg.StartTone(divisor, duration_ms);
}



#define Tone_Update(); {}			// For backwards compatibility.

/*
  // Stop the note if its completion time has come.
void Tone_Update(void)
{
 // Obsolete with current version of library; sounds stop automatically.
 // please remove this function from your program if it is there.
}
*/


void EditColor(byte WhichColor, byte RedComponent, byte GreenComponent, byte BlueComponent)
{
   ColorTable[WhichColor][0] = RedComponent;
   ColorTable[WhichColor][1] = GreenComponent;
   ColorTable[WhichColor][2] = BlueComponent;
 } 




void MeggyJrSimpleSetup(void) 
  {
		Meg = MeggyJr();  
		
	    lastButtonState = Meg.GetButtons();
		
		 Meg.StartTone(0, 0);
	//	 Tone_Update();
		 SoundOff();
  }
  
  

MeggyJr.h
/*
  MeggyJr.h  
 
 Part of the Meggy Jr RGB library for Arduino
 Version 1.5 - 12/31/2011
 Copyright (c) 2011 Windell H. Oskay.  All right reserved.
 
 This library is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this library.  If not, see <http://www.gnu.org/licenses/>.
 
 Thanks to Arthur J. Dahm III and Jay Clegg for code written for the Peggy 2.0,
 which was adapted to make this library.  
 
 */




// If the white balance on your Meggy Jr RGB is significantly off-- it looks purple --
// it may be that you have an LED display with a slightly different green LED elements.
// If that is the case, comment out the following line,
// (Changing "#define UseNewColors  1"  to "//#define UseNewColors  1") 
// so that your sketches will automatically use the old color map:

#define UseNewColors  1


// If it STILL doesn't look good, uncomment the following line:

//#define UseColorMap3


#ifndef byte
#define byte uint8_t
#endif

#ifndef MeggyJr_h
#define MeggyJr_h



#include <inttypes.h>
#include <avr/interrupt.h>
#include <avr/io.h>

#define FPS 120						// Frames per second; value can be changed somewhat.
#define DISP_BUFFER_SIZE 192		// 8 rows * 24 bytes per row == 192. Don't mess with it. :)
#define MAX_BRIGHTNESS 15			// 16 steps of brightness per pixel



// Predefined Colors: 

#ifdef UseColorMap3


#define MeggyDark      0,  0,   0 
#define MeggyRed       12,  0,   0 
#define MeggyOrange   12,  1,   0 
#define MeggyYellow    10,  4,  0
#define MeggyGreen     0,  5,  0
#define MeggyBlue      0,   0,  5
#define MeggyViolet    8,   0,  4
#define MeggyWhite     14,  4,  2 

#define MeggyDimRed    4,  0,   0
#define MeggyDimGreen  0,  1,   0
#define MeggyDimBlue   0,  0,   1
#define MeggyDimOrange 6,  1,   0 
#define MeggydimYellow 4,  1,   0
#define MeggydimAqua   0,  3,   1
#define MeggydimViolet 2,  0,   1


#else

#ifdef UseNewColors

#define MeggyDark      0,  0,   0 
#define MeggyRed       6,  0,   0 
#define MeggyOrange   12,  1,   0 
#define MeggyYellow    10,  4,  0
#define MeggyGreen     0,  6,  0
#define MeggyBlue      0,   0,  5
#define MeggyViolet    8,   0,  4
#define MeggyWhite     7,  4,  2 

#define MeggyDimRed    2,  0,   0
#define MeggyDimGreen  0,  1,   0
#define MeggyDimBlue   0,  0,   1
#define MeggyDimOrange 5,  1,   0 
#define MeggydimYellow 3,  1,   0
#define MeggydimAqua   0,  3,   1
#define MeggydimViolet 2,  0,   1

#else  

#define MeggyDark      0,  0,   0
#define MeggyRed       6,  0,   0
#define MeggyOrange   12,  5,   0
#define MeggyYellow    7,  10,  0
#define MeggyGreen     0,  15,  0
#define MeggyBlue      0,   0,  5
#define MeggyViolet    8,   0,  4
#define MeggyWhite     3,  15,  2 

#define MeggyDimRed    1,  0,   0
#define MeggyDimGreen  0,  1,   0
#define MeggyDimBlue   0,  0,   1
#define MeggyDimOrange 1,  1,   0 
#define MeggydimYellow 1,  1,   0
#define MeggydimAqua   0,  3,   1
#define MeggydimViolet 2,  0,   1

#endif
#endif

//#define SoundEnabled (TCCR1B > 0)

// Optional timing test routine; can be used to measure
// duration of interrupt refresh routine.
//#define  timingtest 1


class MeggyJr
{
	public:
		MeggyJr(void);										// Constructor
		
		static byte MeggyFrame[DISP_BUFFER_SIZE];
		static byte AuxLEDs;
		
		static byte SoundEnabled;		static byte SoundAllowed;	 
		
	//	 static volatile byte ToneInProgress;		// Removed, v 1.3		 
	//	static unsigned long SoundStopTime;			// Removed, v 1.3
		
		void ClearMeggy (void);	
		void ClearPixel(byte x, byte y);
		byte GetButtons(void);

		byte GetPixelR(byte x, byte y);
		byte GetPixelG(byte x, byte y);
		byte GetPixelB(byte x, byte y);
		 
		void SetPxClr(byte x, byte y, byte rgb[3]);
		static byte *currentColPtr;
		static byte  currentCol;
		static byte  currentBrightness;

		void SoundState(byte t);
        void StartTone(unsigned int Tone, unsigned int duration)  ;  
//		void SoundCheck(void) ;  // End sound if it's time to do so.   // Removed, v 1.3

 		static unsigned int ToneTimeRemaining;			// new, v 1.3

#ifdef timingtest
	static unsigned int testTime;
#endif
	    
	private: 
		
};

#endif

